cparser/Elab.ml

```diff
@@ -983,13 +1002,18 @@ let init_int_array_wstring opt_size s =
 let check_init_type loc env a ty =
   if wrap2 valid_assignment loc env a ty then ()
   else if wrap2 valid_cast loc env a.etyp ty then
-    warning loc
-      "initializer has type@ %a@ instead of the expected type @ %a"
-      Cprint.typ a.etyp Cprint.typ ty
+    if wrap2 int_pointer_conversion loc env a.etyp ty then
+      warning loc Int_conversion
+        "incompatible integer to pointer conversion initializing %a with an expression of type %a"
+        (print_typ env) ty (print_typ env) a.etyp
+    else
+      warning loc Int_conversion
+        "incompatible pointer to integer conversion initializing %a with an expression of type %a"
+        (print_typ env) ty (print_typ env) a.etyp
   else
     error loc
-      "initializer has type@ %a@ instead of the expected type @ %a"
-      Cprint.typ a.etyp Cprint.typ ty
+      "initializing %a with an expression of incompatible type %a"
+      (print_typ env) ty (print_typ env) a.etyp
```

Comment:
This is clang's wording, but it is quite bad:
```
   gee.c:3:5: error: initializing 'int' with an expression of incompatible type 'struct S'
```
It is not `int` that is being initialized, but something (a variable or an array element or a member of a struct/union) that has type `int`.  The original wording avoided this problem.